### ISSUE TITLE
Update constructors

### ISSUE THREAD

**Issue by shyun3** [tcbrindle/span#32]
Since C++20 has been finalized, [cppreference](https://en.cppreference.com/w/cpp/container/span/span) has updated their section on constructors. Looks like (2) and (3) now use generic iterators instead of concrete pointers and the container constructors have been replaced by the range constructor in (7).

**Comment by tcbrindle** [comment]
Agreed, I was holding off on this to see what other changes might emerge after the Prague meeting, but now the design is fixed we should implement it as best we can.

I say "as best we can" because the updated constructors now use concepts, which are [quite a lot of work](github.com/tcbrindle/nanorange) to emulate using SFINAE. I'll have to experiment with how this can be done without introducing too much template goo.
